src/Verify.MSTest/Verifier.cs

```diff
@@ -0,0 +1,215 @@
+        if (context != null)
+        {
+            var fileName = autoVerify ? path.VerifiedPath : path.ReceivedPath;
+            context.AddResultFile(fileName);
+        }
+
+        return Task.CompletedTask;
+    }
+
+    static Verifier()
+    {
+        VerifierSettings.OnFirstVerify((pair, _, autoVerify) => AddFile(pair, autoVerify));
+        VerifierSettings.OnVerifyMismatch((pair, _, autoVerify) => AddFile(pair, autoVerify));
+    }
+
+    public static readonly AsyncLocal<TestContext?> CurrentTestContext = new();
+
+    static InnerVerifier BuildVerifier(VerifySettings settings, string sourceFile, bool useUniqueDirectory)
+    {
+        if (useUniqueDirectory)
+        {
+            settings.UseUniqueDirectory();
+        }
+
+        if (CurrentTestContext.Value is null)
+        {
+            throw new($"TestContext is null. {AttributeUsageHelp}");
+        }
+
+        var testName = CurrentTestContext.Value.TestName;
+        if (testName is null)
+        {
+            throw new($"TestContext.TestName is null. {AttributeUsageHelp}");
+        }
+
+        var typeName = CurrentTestContext.Value.FullyQualifiedTestClassName;
+        if (typeName is null)
+        {
+            throw new($"TestContext.FullyQualifiedTestClassName is null. {AttributeUsageHelp}");
+        }
+
+        if(!TryGetTypeFromTestContext(typeName, CurrentTestContext.Value, out var type))
+        {
+            type = FindType(typeName);
+        }
+
+        var testNameSpan = testName.AsSpan();
+        var indexOf = testNameSpan.IndexOf('(');
+        if (indexOf > 0)
+        {
+            testNameSpan = testNameSpan[..indexOf];
+        }
+
+        indexOf = testNameSpan.IndexOf('.');
+        if (indexOf > 0)
+        {
+            testNameSpan = testNameSpan[(indexOf + 1)..];
+        }
+
+        VerifierSettings.AssignTargetAssembly(type.Assembly);
+        var method = FindMethod(type, testNameSpan);
+
+        var pathInfo = GetPathInfo(sourceFile, type, method);
+        return new(
+            sourceFile,
+            settings,
+            type.NameWithParent(),
+            method.Name,
+            method.ParameterNames(),
+            pathInfo);
+    }
+
+    /// <summary>
+    /// As an optimization, try to retrieve the <see cref="ITestMethod"/> stored on the
+    /// <see cref="TestContext"/>, and use that to retrieve the <see cref="Type"/>.
+    ///
+    /// If reflection fails, return <c>false</c>.
+    /// </summary>
+    /// <param name="typeName">The fully qualified name of the class of the currently running test.</param>
+    /// <param name="testContext">The <see cref="TestContext"/> of the current test.</param>
+    /// <param name="type">The <see cref="Type"/> of the currently running test.</param>
+    /// <returns><c>true</c> if the reflection succeeded; <c>false</c> otherwise.</returns>
+    static bool TryGetTypeFromTestContext(string typeName, TestContext testContext, [NotNullWhen(true)] out Type? type)
```

Comment:
why so defensive in this method. IMO it should assume the correct shape and throw if that shape is not correct

Reply:
I'm hesitant to assert on the specific shape here because I'm reaching into the internals of MSTest and not relying on any public API contracts.

More specifically, I'm concerned that this code would break and become an adoption blocker for early adopters of [TestAnywhere](https://github.com/microsoft/testfx/blob/main/docs/testingplatform/Index.md) (I'm not a user of it myself, but I did verify this code works with the current version).

In general I'm not super thrilled with the amount of reflection going on here, but I think reducing the amount would require digging more into Verify's internals, which feels to me like it should be a separate change.

Thus, I'd prefer leaving it as-is, but I'll do whatever you'd like. I'll volunteer to do follow up work in this space to reduce the amount of reflection overall, but I'd need some direction or a brain dump from you on how to proceed.

Let me know if this makes sense. Thanks!

Reply:
An idea that just hit me would be to augment what we stash in the AsyncLocal.

Rather than try to get back to the test class after-the-fact, grab everything we need in the TestContext property setter (which means we're already in the test class). Then `Verifier` or `VerifyBase` doesn't need to do any work.

If possible I'd like to keep this PR source generator focused, so if this sounds good to you I can do that as a follow up change, or even as a standalone change and then merge that into this branch.

Thoughts?

Reply:
> An idea that just hit me would be to augment what we stash in the AsyncLocal.

sounds good. i will merge this now. can u have a go at an AsyncLocal PR next?